### PR TITLE
extend `nonzero` to int64

### LINKED ISSUE #51871: Tensor.nonzero fails on GPU for tensors containing more than INT_MAX elements
## 🐛 Bug

Calling nonzero for a tensor with more than INT_MAX elements leads to an error on GPU. On CPU there does not seem to be a problem.

## To Reproduce

```python-console
>>> x = torch.zeros(2**20, 2**11, device="cuda")
>>> x.nonzero()
Traceback (most recent call last):
  File "<input>", line 1, in <module>
RuntimeError: nonzero is not supported for tensors with more than INT_MAX elements,   file a support request
```
## Expected behavior

Return the indices of non-zero elements.

## Environment

```
Collecting environment information...
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 450.102.04
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```

## Additional context

The problem was already described on [discuss.pytorch.org](https://discuss.pytorch.org/t/torch-geometric/106975), but I could not find the corresponding issue.

cc @ngimel

### PR DESCRIPTION
**Pull request by bhack**
Fixes #51871

### CODE REVIEW

**Comment by bhack** [comment]
/cc @ezyang @eqy  This is an explorative blackbox PR as I don't have free cuda resources right now and we don't have a quick way to setup the env to contribute sparse c++/cuda PR (see https://github.com/pytorch/pytorch/issues/125297).

But I made this editable on your side in the case you have the env ready and a quick fix is enough.

**Review by eqy — COMMENTED** [review]
Should we add a (presumably large tensor) test for this?

**Comment by bhack** [comment]
> Should we add a (presumably large tensor) test for this?

Do we had an INT_MAX test already somewhere that we could expand?

**Comment by eqy** [comment]
Unfortunately these are not really unified at the moment, but this should surface some examples: https://github.com/search?q=repo%3Apytorch%2Fpytorch+64bit+language%3APython+path%3A%2F%5Etest%5C%2F%2F&type=code

**Comment by bhack** [comment]
As we don't have a specific CUDA test do we want to find a workaround from python?

Can you suggest one from `grep -R torch.nonzero test/`?

**Comment by bhack** [comment]
I think I am going to close this as `cub::DispatchSelectIf` probably it will be slower then `cub::DeviceSelect::Flagged` we are currently using.

Probably we need to wait upstream for https://github.com/NVIDIA/cccl/issues/1422

What do you think?

**Comment by bhack** [comment]
@ezyang Do you think we can we open a new ticket to lower this with Trition `when` and `sum`?

File: aten/src/ATen/native/cuda/Nonzero.cu [hunk 1]
```diff
@@ -18,19 +18,60 @@
...
+  OutputIteratorT d_out,
+  NumSelectedIteratorT d_num_selected_out,
+  OffsetT num_items,
+  SelectOp select_op,
+  cudaStream_t stream = 0)
+{
+    using flag_iterator_t = cub::NullType*;
+    using equality_op_t   = cub::NullType;
+
+    return cub::DispatchSelectIf<
```
Comment (bhack) [inline aten/src/ATen/native/cuda/Nonzero.cu:43]: Do you we need to wait for https://github.com/pytorch/pytorch/pull/125997 for a compatible API?

Comment (bhack) [reply aten/src/ATen/native/cuda/Nonzero.cu:43]: Or cub/cccl 2.4.0?

**Comment by ezyang** [comment]
yes, need a big tensor test. @eqy's link is good for examples

**Comment by bhack** [comment]
Ok thanks,
so I am going to close it as I don't have the env and currently spare GPU computing to write a brand new test and recompile it.
At least if we don't identify another python test that it is already indirectly using nonzero and  that we could modify it with a big input.

**Comment by bhack** [comment]
Just to check if it could compile at least with the current CUB version.
Do you know what is this CI failure
```cuda
/usr/local/cuda/include/cub/agent/agent_select_if.cuh(264): error: function "at::native::<unnamed>::NonZeroOp<T>::operator() [with T=c10::complex<c10::Half>]" cannot be called with the given argument list
            argument types are: (int64_t)
            object type is: at::native::<unnamed>::NonZeroOp<c10::complex<c10::Half>>
                  selection_flags[ITEM] = select_op(items[ITEM]);
```

**Comment by bhack** [comment]
I think we need `cub/cub/agent/agent_select_if.cuh` changes introduced at https://github.com/NVIDIA/cccl/pull/1379

So this mean that we need to wait for the next cuda 12.4 update and make it also conditional.

**Comment by ezyang** [comment]
This PR seems fine. I agree you may need to preprocessor your way to victory. CI will say.

**Comment by ezyang** [comment]
if you're willing to wait for cuda 12.5 :)

**Comment by bhack** [comment]
> if you're willing to wait for cuda 12.5 :)

This version is required for the workaround API. A full upstream solution it will require to wait more CUDA releases.

**Comment by ezyang** [comment]
Sorry, I'm not sure I understand the state of this PR. I would happily accept a PR that makes nonzero int64 work on CUDA 12.5 or later, subject to the requirement that this functionality is preprocessored out. If you don't mind waiting, we can also ice this PR until CUDA 12.6 shows up by default and then we can just land it as is.

**Comment by bhack** [comment]
The current status for testing/using the workaround in this PR is to have `CUB 2.4.0`. 
As in pytorch we currently use CUB from the official CUDA distribution it means we need upgrade pytorch to support CUDA 12.5 (any plan?).

For a "regular" upstream solution we need that this will be merged:
https://github.com/NVIDIA/cccl/issues/1422

And after that we need an official CUB release and that release will be included in a CUDA release. 
So we have no guaranties that we will have an upstream solution with CUDA 12.6

**Comment by bhack** [comment]
@ezyang See https://github.com/NVIDIA/cccl/issues/1454#issuecomment-2138705800

**Comment by ezyang** [comment]
@atalman for CUDA 12.5

**Comment by bhack** [comment]
We are waiting pytorch adopting CUDA 12.5

**Comment by bhack** [comment]
@ezyang In case you're interested, it appears that META's release of SAMv2 yesterday might be affected by this issue: https://github.com/facebookresearch/segment-anything-2/issues/44.

**Comment by bhack** [comment]
the bot does not forgive...

**Pull request closed without merging**